src/dmd/expressionsem.d

```diff
@@ -5764,6 +5764,161 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
         exp.e1 = exp.e1.optimize(WANTvalue);
         exp.e1 = exp.e1.toBoolean(sc);
 
+        enum NegatedState { none, equals, notEquals }
+        /*
+        DMD lowers struct equality comparison for complex fields into a
+        conjunctive form of their individual fields.
+
+        Example:
+            S  { int a; float f;}
+            S s1, s2;
+            s1 == s2;
+            ->
+            s1.a == s2.a && s1.f == s2.f
+
+        Here, we try to conservatively detect such a pattern and return the respective
+        structs as `left` and `right`.
```

Comment:
This approach appears to be:

1. compiler converts == to a conjunctive form
2. this code detects a conjunctive form and more or less reverts it back to the original

This seems like a very fragile and backwards approach. It's fragile in that it's highly dependent on exactly what happens with earlier lowerings, and fragile because it can falsely detect patterns written directly by the user.

There's got to be a better way. It's also unclear to me if this PR a problem that even needs a solution.

Reply:
>  reverts it back to the original

Well, it doesn't revert it. It just adds the detected structs to `_d_assertfail`, s.t. we can actually get a sensible context message with `-checkaction=context`.

>  It's also unclear to me if this PR a problem that even needs a solution.

`-checkaction=context` is an _existing_ switch that rewrites the assert message, s.t. the user sees the error message directly in the error and doesn't have to rerun the program. Apart from generally improving run/debug ergonomics, this is very important if you have a testsuite that takes a long time or runs on a remote server (e.g. CI or production) where it's hard to reproduce the failure.

tl;dr: with `-checkaction=context` you just get:

```
core.exception.AssertError@foo.d(14): Assertion failure
```

More detailed explanation: https://dlang.org/changelog/2.085.0.html#assert

> This seems like a very fragile and backwards approach. It's fragile in that it's highly dependent on exactly what happens with earlier lowerings,

Yup, but I don't know a better way to do it in DMD.
Also, if these lowerings get ever changed, it would just fallback back to old "assert failed XYZ" where it just stringifies the AST.

> and fragile because it can falsely detect patterns written directly by the user.

No, it won't have any false positives with this approach as it only adds the structs to the assert message if every part of the expression consists of binary expressions of both of them. 
A few examples of how DMD behaves _currently_ with `-checkaction=context`:

1) Field comparison

```
assert(t1.a == t2.a);
->
core.exception.AssertError@foo.d(14): 0 != 2
```

2) Struct comparison

```
assert(t1 == t2);
->
assert(t1 == t2) failed
```

3) User-generated struct comparison

```
assert(t1.a == t2.a && t1.f == t2.f);
->
assert(t1.a == t2.a && (t1.f == t2.f)) failed
```

Now with this PR (2) would result in e.g.:

```diff
-assert(t1 == t2) failed
+T2(a:0, f:nan, s:"") != T2(a:2, f:nan, s:"")
```

Note that the code currently doesn't even rewrite (3), though if it did that would even be an improvement.

> There's got to be a better way

I'm all open for better solutions.